File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Scenarios/RcB2ProtobufCodecScenario.java
package systems.zlink.e2e.registrationcodec.client.Scenarios;

import systems.zlink.e2e.registrationcodec.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrationcodec.client.Support.ScenarioContext;
import systems.zlink.e2e.registrationcodec.shared.Contracts;

public final class RcB2ProtobufCodecScenario {
    private RcB2ProtobufCodecScenario() {
    }

    public static void run(ScenarioContext context) {
        String protobufValue = context.server()
            .post("/codec/roundtrip")
            .submit(Contracts.CodecScenarioRes.class).toCompletableFuture().join().body()
            .protobufValue();
        ScenarioAssert.ensure("echo:protobuf-request".equals(protobufValue),
            "RC-B2 request mismatch");
        ScenarioAssert.waitForEvidence(
            context.evidence(),
            "ContentType",
            "ProtobufEcho",
            "application/x-protobuf");
        ScenarioAssert.waitForEvidence(context.evidence(), "Send", "ProtobufEcho", "protobuf-send");
        System.out.println("scenario RC-B2 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/client/Scenarios/RcB2ProtobufCodecScenario.kt
package systems.zlink.e2e.kotlin.registrationcodec.client.scenarios

import systems.zlink.e2e.kotlin.registrationcodec.client.support.CodecScenarioResult
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrationcodec.client.support.ScenarioHttpClient

class RcB2ProtobufCodecScenario(
    private val server: ScenarioHttpClient,
    private val assert: ScenarioAssert,
) {
    fun run(): CodecScenarioResult {
        val protobufReply = server.post("/codec/protobuf", null, Map::class.java)
        val value = protobufReply["value"] as String
        assert.that(value == "echo:protobuf-request", "RC-B2 request mismatch")
        assert.that(
            server.waitForEvidence("Send", "ProtobufEcho", "protobuf-send")
                .any { it.marker == "Send" && it.packetName == "ProtobufEcho" && it.value == "protobuf-send" },
            "RC-B2 send evidence missing",
        )
        println("scenario RC-B2 passed")
        return CodecScenarioResult("RC-B2", "ProtobufEcho", value)
    }
}
